File: src/common/platform.cpp
// ============================================================
// 跨平台Socket封装 - 实现
// ============================================================

#include "common/platform.h"
#include <sstream>
#include <vector>
#include <utility>

#ifdef _WIN32
    #pragma comment(lib, "ws2_32.lib")
#endif

namespace NetworkUtils {

// ----------------------------------------------------------
// 初始化网络库
// ----------------------------------------------------------
bool initialize() {
#ifdef _WIN32
    // Windows: 必须调用 WSAStartup 初始化 Winsock DLL
    WSADATA wsa_data;
    int result = WSAStartup(MAKEWORD(2, 2), &wsa_data);
    if (result != 0) {
        return false;
    }
#endif
    return true;
}

// ----------------------------------------------------------
// 清理网络库
// ----------------------------------------------------------
void cleanup() {
#ifdef _WIN32
    // Windows: 清理 Winsock
    WSACleanup();
#endif
}

// ----------------------------------------------------------
// 获取错误描述字符串
// ----------------------------------------------------------
std::string get_error_string(int error_code) {
#ifdef _WIN32
    LPVOID msg_buf = nullptr;
    FormatMessageA(
        FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
        nullptr, error_code,
        MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
        (LPSTR)&msg_buf, 0, nullptr
    );
    if (msg_buf) {
        std::string msg((char*)msg_buf);
        LocalFree(msg_buf);
        // 去除尾部换行符
        while (!msg.empty() && (msg.back() == '\n' || msg.back() == '\r')) {
            msg.pop_back();
        }
        return msg;
    }
    return "Unknown error: " + std::to_string(error_code);
#else
    return std::string(strerror(error_code));
#endif
}

/**
 * @brief 获取最后一次Socket错误的描述信息
 *
 * @details
 * 在 Windows 上获取 WSAGetLastError() 的结果
 * 在 Linux/macOS 上获取 errno 的当前值
 *
 * @return std::string 错误信息字符串，包含错误码和描述
 */
std::string get_last_error_string() {
    int err = GET_SOCKET_ERROR();
    return "[错误码 " + std::to_string(err) + "] " + get_error_string(err);
}

// ----------------------------------------------------------
// 判断是否为"操作会阻塞"错误
// ----------------------------------------------------------
bool is_would_block(int error_code) {
#ifdef _WIN32
    return error_code == WSAEWOULDBLOCK;
#else
    // EINPROGRESS: 非阻塞 connect() 正在建立连接 (正常状态!)
    // EWOULDBLOCK / EAGAIN: 非阻塞 read/write 时缓冲区无数据/已满
    return error_code == EWOULDBLOCK || error_code == EAGAIN || error_code == EINPROGRESS;
#endif
}

// ----------------------------------------------------------
// 设置Socket为非阻塞模式
// ----------------------------------------------------------
bool set_nonblocking(SOCKET_FD sock) {
#ifdef _WIN32
    u_long mode = 1;
    return ioctlsocket(sock, FIONBIO, &mode) == 0;
#else
    int flags = fcntl(sock, F_GETFL, 0);
    if (flags == -1) return false;
    return fcntl(sock, F_SETFL, flags | O_NONBLOCK) == 0;
#endif
}

// ----------------------------------------------------------
// 设置Socket为阻塞模式
// ----------------------------------------------------------
bool set_blocking(SOCKET_FD sock) {
#ifdef _WIN32
    u_long mode = 0;
    return ioctlsocket(sock, FIONBIO, &mode) == 0;
#else
    int flags = fcntl(sock, F_GETFL, 0);
    if (flags == -1) return false;
    return fcntl(sock, F_SETFL, flags & ~O_NONBLOCK) == 0;
#endif
}

// ----------------------------------------------------------
// 设置Socket地址重用
// ----------------------------------------------------------
bool set_reuse_addr(SOCKET_FD sock) {
    int optval = 1;
#ifdef _WIN32
    return setsockopt(sock, SOL_SOCKET, SO_REUSEADDR,
                      (const char*)&optval, sizeof(optval)) == 0;
#else
    return setsockopt(sock, SOL_SOCKET, SO_REUSEADDR,
                      &optval, sizeof(optval)) == 0;
#endif
}

// ----------------------------------------------------------
// 获取本机所有IPv4地址
// ----------------------------------------------------------
std::vector<std::string> get_local_ips() {
    std::vector<std::string> ips;

#ifdef _WIN32
    // Windows: 使用 getaddrinfo 获取本机地址
    char hostname[256];
    if (gethostname(hostname, sizeof(hostname)) != 0) {
        return ips;
    }

    struct addrinfo hints, *result = nullptr;
    memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_INET;        // 仅IPv4
    hints.ai_socktype = SOCK_STREAM;
    hints.ai_flags = AI_PASSIVE;

    if (getaddrinfo(hostname, nullptr, &hints, &result) != 0) {
        return ips;
    }

    for (struct addrinfo* ptr = result; ptr != nullptr; ptr = ptr->ai_next) {
        struct sockaddr_in* addr = (struct sockaddr_in*)ptr->ai_addr;
        char ip_str[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, &addr->sin_addr, ip_str, sizeof(ip_str));
        ips.push_back(ip_str);
    }
    freeaddrinfo(result);

#else
    // Linux/macOS: 遍历网络接口获取IPv4地址
    struct ifaddrs* ifaddr;
    if (getifaddrs(&ifaddr) == -1) {
        return ips;
    }

    for (struct ifaddrs* ifa = ifaddr; ifa != nullptr; ifa = ifa->ifa_next) {
        if (ifa->ifa_addr == nullptr) continue;
        // 只处理IPv4地址
        if (ifa->ifa_addr->sa_family != AF_INET) continue;
        // 跳过回环地址127.0.0.1
        struct sockaddr_in* addr = (struct sockaddr_in*)ifa->ifa_addr;
        char ip_str[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, &addr->sin_addr, ip_str, sizeof(ip_str));
        std::string ip(ip_str);
        if (ip != "127.0.0.1") {
            ips.push_back(ip);
        }
    }
    freeifaddrs(ifaddr);
#endif

    return ips;
}

// ----------------------------------------------------------
// 获取本机IPv4地址和子网掩码
// ----------------------------------------------------------
std::vector<std::pair<std::string, std::string>> get_local_ips_with_mask() {
    std::vector<std::pair<std::string, std::string>> result;

#ifdef _WIN32
    // Windows: 简单返回没有子网掩码的IP
    for (const auto& ip : get_local_ips()) {
        result.emplace_back(ip, "255.255.255.0");  // 默认 /24
    }
#else
    struct ifaddrs* ifaddr;
    if (getifaddrs(&ifaddr) == -1) {
        return result;
    }

    for (struct ifaddrs* ifa = ifaddr; ifa != nullptr; ifa = ifa->ifa_next) {
        if (ifa->ifa_addr == nullptr || ifa->ifa_netmask == nullptr) continue;
        if (ifa->ifa_addr->sa_family != AF_INET) continue;

        struct sockaddr_in* addr = (struct sockaddr_in*)ifa->ifa_addr;
        struct sockaddr_in* mask = (struct sockaddr_in*)ifa->ifa_netmask;

        char ip_str[INET_ADDRSTRLEN];
        char mask_str[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, &addr->sin_addr, ip_str, sizeof(ip_str));
        inet_ntop(AF_INET, &mask->sin_addr, mask_str, sizeof(mask_str));

        std::string ip(ip_str);
        if (ip != "127.0.0.1") {
            result.emplace_back(ip, std::string(mask_str));
        }
    }
    freeifaddrs(ifaddr);
#endif

    return result;
}

} // namespace NetworkUtils

File: python/p2p_transfer/common/platform.py
"""
平台相关网络工具模块

提供跨平台的底层网络操作：
- 本地 IP 地址枚举（netifaces / ip 命令回退）
- 主网卡 IP 选择（UDP connect 技巧）
- 非阻塞/阻塞模式切换
- SO_REUSEADDR 设置
- BBR 拥塞控制尝试启用（Linux 内核 4.9+）
- 可恢复的网络错误判断（EAGAIN/EWOULDBLOCK/EINPROGRESS）
"""

import socket
import struct
import errno
import fcntl
import os
from .types import MULTICAST_ADDR, SIGNALING_PORT


# 可恢复的网络错误码集合：
# EWOULDBLOCK - 非阻塞套接字无数据可读/写
# EAGAIN - 同 EWOULDBLOCK（Linux 下两者等值）
# EINPROGRESS - 非阻塞 connect() 正在进行中（必须纳入判断，否则 connect 会误判为失败）
SOCKET_ERR = (
    errno.EWOULDBLOCK,
    errno.EAGAIN,
    errno.EINPROGRESS,
)


def get_local_ips() -> list[str]:
    """获取本机所有非回环 IPv4 地址

    优先使用 netifaces 库，若未安装则调用 ip 命令解析输出，
    最终回退到 127.0.0.1。
    """
    ips = []
    # 方案 1：使用 netifaces 库（可靠的跨平台方案）
    try:
        import netifaces
        for iface in netifaces.interfaces():
            addrs = netifaces.ifaddresses(iface)
            if netifaces.AF_INET in addrs:
                for addr in addrs[netifaces.AF_INET]:
                    ip = addr.get("addr", "")
                    if ip and not ip.startswith("127."):
                        ips.append(ip)
        return ips if ips else ["127.0.0.1"]
    except ImportError:
        pass

    # 方案 2：调用 ip 命令（Linux 专用，带 3 秒超时防止卡死）
    try:
        import subprocess
        import re
        output = subprocess.check_output(
            ["ip", "-4", "addr", "show"], text=True, timeout=3
        )
        for line in output.splitlines():
            m = re.search(r"inet\s+(\d+\.\d+\.\d+\.\d+)", line)
            if m:
                ip = m.group(1)
                if not ip.startswith("127."):
                    ips.append(ip)
        if ips:
            return ips
    except Exception:
        pass

    # 方案 3：全部回退无效，返回回环地址
    return ["127.0.0.1"]


def select_local_ip() -> str:
    """选择本机主网卡 IPv4 地址

    使用 UDP connect 技巧：创建一个临时的 UDP 套接字并尝试连接到
    1.1.1.1:53（Cloudflare DNS，不实际发送数据），然后通过 getsockname()
    获取内核路由选择出的源 IP，即为主网卡 IP。
    失败时回退到 get_local_ips() 的第一个结果。
    """
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(("1.1.1.1", 53))          # 触发内核路由选择，不发送实际 UDP 包
        ip = s.getsockname()[0]
        s.close()
        if ip and ip != "0.0.0.0":
            return ip
    except:
        pass

    ips = get_local_ips()
    return ips[0] if ips else "127.0.0.1"


def set_nonblocking(sock: socket.socket):
    """将套接字设置为非阻塞模式（O_NONBLOCK）"""
    fd = sock.fileno()
    flags = fcntl.fcntl(fd, fcntl.F_GETFL)
    fcntl.fcntl(fd, fcntl.F_SETFL, flags | os.O_NONBLOCK)


def set_blocking(sock: socket.socket):
    """将套接字恢复为阻塞模式（清除 O_NONBLOCK 标志）"""
    fd = sock.fileno()
    flags = fcntl.fcntl(fd, fcntl.F_GETFL)
    fcntl.fcntl(fd, fcntl.F_SETFL, flags & ~os.O_NONBLOCK)


def set_reuse_addr(sock: socket.socket):
    """设置 SO_REUSEADDR 选项，允许端口复用和快速重启"""
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)


def enable_bbr(sock: socket.socket) -> bool:
    """尝试为 TCP 套接字启用 BBR 拥塞控制算法

    TCP_CONGESTION = 13（Linux 内核常量）。
    需要 Linux 内核 4.9+，若系统不支持则静默回退到默认算法（通常 cubic）。
    """
    try:
        sock.setsockopt(socket.IPPROTO_TCP, 13, b"bbr")
        return True
    except (OSError, AttributeError):
        return False


def is_would_block(err: int) -> bool:
    """判断 errno 值是否为可恢复的"会阻塞"错误

    包括 EWOULDBLOCK、EAGAIN（资源暂时不可用）和 EINPROGRESS（连接中）。
    用于非阻塞 socket 操作后的错误处理。
    """
    return err in SOCKET_ERR
